### PR TITLE
Dialogue correctly handles optional binary responses

### PR DESCRIPTION
**Pull request by carterkozak**
This allows the generator to consume changes from
https://github.com/palantir/dialogue/pull/501

## After this PR
==COMMIT_MSG==
Dialogue correctly handles optional binary responses
==COMMIT_MSG==

### CODE REVIEW

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java [hunk 1]
```diff
@@ -132,8 +132,8 @@ public MethodSpec generate(ServiceDefinition def) {
     }
 
     private Optional<FieldSpec> deserializer(EndpointName endpointName, Optional<Type> type) {
-        // TODO(forozco): handle optional binary ??
-        if (type.map(t -> t.accept(TypeVisitor.IS_BINARY)).orElse(false)) {
+        if (type.map(t -> t.accept(TypeVisitor.IS_BINARY) || isOptionalBinary(t))
```
Comment (ferozco) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java]: can we factor this lambda out into a separate function?

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java [hunk 2]
```diff
@@ -152,6 +152,12 @@ public MethodSpec generate(ServiceDefinition def) {
                 .build());
     }
 
+    // TODO(ckozak): Type should be dealiased
```
Comment (ferozco) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java]: we basically have an alias resolving visitor in this class already so might as well implement

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java (hunk 1 above)
Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java]: Extracted this with support for dealiasing.

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java (hunk 2 above)
Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java]: :+1: updated

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/DialogueServiceGenerator.java [hunk 3]
```diff
@@ -49,10 +49,14 @@ public DialogueServiceGenerator(Options options, String apiVersion) {
 
         TypeMapper parameterTypes = new TypeMapper(
                 conjureDefinition.getTypes(),
-                new ClassVisitor(conjureDefinition.getTypes(), options, ClassVisitor.Mode.PARAMETER));
+                new DialogueClassNameVisitor(
```
Comment (ferozco) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/DialogueServiceGenerator.java]: its a bit confusing to have a `DialogueClassNameVisitor` and a `ClassVisitor` both of which encode dialogue specific details. could we unify them in some way?

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/DialogueServiceGenerator.java]: Ya, I didn't grok that ClassVisitor was dialogue specific, I can combine the two. Mind if I rename it to include dialogue?

Comment (ferozco) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/DialogueServiceGenerator.java]: please do!

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/DialogueServiceGenerator.java]: Should be good to go

**Comment by ferozco** [comment]
👍

**Pull request merged**